cdap-common/src/main/java/co/cask/cdap/common/guice/KafkaClientModule.java

```diff
@@ -44,9 +54,56 @@ protected void configure() {
   @Provides
   @Singleton
   private KafkaClientService providesKafkaClientService(CConfiguration cConf, ZKClient zkClient) {
+    if (cConf.get(KafkaConstants.ConfigKeys.ZOOKEEPER_QUORUM) != null) {
+      return new DefaultKafkaClientService(cConf);
+    }
     String kafkaZKNamespace = cConf.get(KafkaConstants.ConfigKeys.ZOOKEEPER_NAMESPACE_CONFIG);
     return new ZKKafkaClientService(
       kafkaZKNamespace == null ? zkClient : ZKClients.namespace(zkClient, "/" + kafkaZKNamespace)
     );
   }
+
+  private static final class DefaultKafkaClientService extends AbstractIdleService implements KafkaClientService {
+
+    private final ZKClientService zkClientService;
+    private final KafkaClientService delegate;
+
+    DefaultKafkaClientService(CConfiguration cConf) {
+      zkClientService = ZKClientServices.delegate(
+        ZKClients.reWatchOnExpire(
+          ZKClients.retryOnFailure(
+            ZKClientService.Builder.of(cConf.get(KafkaConstants.ConfigKeys.ZOOKEEPER_QUORUM))
+              .setSessionTimeout(cConf.getInt(Constants.Zookeeper.CFG_SESSION_TIMEOUT_MILLIS,
+                                              Constants.Zookeeper.DEFAULT_SESSION_TIMEOUT_MILLIS))
+              .build(),
+            RetryStrategies.exponentialDelay(500, 2000, TimeUnit.MILLISECONDS)
+          )
+        )
+      );
+      delegate = new ZKKafkaClientService(zkClientService);
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+      zkClientService.startAndWait();
+      delegate.startAndWait();
+
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
+      delegate.stopAndWait();
+      zkClientService.stopAndWait();
```

Comment:
similarly in here. If delegate stop failed, you still need to stop the zkClientService.